### PR TITLE
feat(lsp): smart paste — lex/preparePaste request + re-anchor transform

### LINKED ISSUE #708: Smart paste: lex/preparePaste request + re-anchor transform (lex-lsp)
Implement smart paste entirely in lex-lsp, behind a new custom request. Editors will add capture-and-apply glue in follow-up PRs (one per editor); this issue is the whole server side and ships first, fully tested.

Spec: `comms/specs/proposals/smart-paste.lex` (comms#73).

## Scope

- **Custom request `lex/preparePaste`** — params `{ textDocument, range, pastedText }`, response `{ text, mode }`. Reuses the already-parsed buffer state; no re-parse from scratch. Pure w.r.t. document state.
- **Capability advertisement** in the initialize response so editors enable interception only against a server that implements it.
- **Paste-mode classification** (§3): `passthrough — verbatim`, `passthrough — table`, `passthrough — single line`, `re-anchor`. Resolved innermost-first.
- **Re-anchor transform** (§4):
  - Anchor = content indentation of the structural container enclosing the range start (not line whitespace, not caret column).
  - Baseline = min leading-whitespace width over non-blank clipboard lines.
  - `delta = anchor - baseline`; per line emit `max(0, original_indent + delta)`, blank lines stay empty.
  - First-line handling: fresh-line case re-anchors every line; merge case strips line 1's whitespace and re-anchors lines 2..n.
- **Tab width**: measure leading whitespace in display columns at Lex tab width (4); emit spaces.

## Tests

Table-driven over the §6 edge cases at minimum:
- empty clipboard, single-line, selection-replace (anchor from selection start)
- positive / negative / zero delta
- fresh-line vs. merge first-line
- verbatim passthrough, table passthrough
- clipboard that is itself a verbatim block (offset keeps body well-formed)
- trailing newline preserved, mixed tabs/spaces, paste at document start, partial indentation

## Out of scope (this issue)

Editor glue (separate PRs), table-aware cell paste, structural/re-parse paste, cross-format paste — all per spec §8.

Part of the smart-paste epic.

### PR DESCRIPTION
**Pull request by arthur-debert**
Implements the server side of smart paste — **lex-fmt/lex#708** (design: comms#73, `specs/proposals/smart-paste.lex`). Editor glue is out of scope and ships as separate per-editor PRs.

## What this does

Lex encodes structure as indentation, so pasting carries the clipboard's *absolute* indentation and lands wrong. A new custom LSP request re-anchors pasted text to the caret's structural context at paste time. One implementation in the server, tested once, serves every editor.

## Changes

- **`lex_lsp_core::prepare_paste`** (new, pure module — "logic over the AST" belongs in the sync core shared by `lexd-lsp` and `lex-wasm`):
  - **Classification (§3)**, innermost-first: `passthrough-verbatim` / `passthrough-table` / `passthrough-single-line` vs. `re-anchor`. Verbatim is parse-driven (`find_verbatim_at_position`) and wins outright; table is the leading-`|` row heuristic.
  - **Anchor (§4.1)**: content indentation of the *enclosing structural container* (session / definition / list item), walked from the parse — **not** the caret line's whitespace. A caret on a container's head line resolves to the parent (anchor is the body indent a new child would carry); a blank line left at column zero deep in a session still anchors correctly.
  - **Transform (§4.2–4.4)**: baseline = min leading-whitespace over non-blank lines; `delta = anchor - baseline`; per line `max(0, original_indent + delta)` spaces; blanks stay empty; fresh-line re-anchors every line, merge strips line 1 and re-anchors the rest. Tab width 4, measured in display columns, emitted as spaces. Trailing newline preserved.
- **`LexLanguageServer::prepare_paste`** — thin tower-lsp wrapper, registered via `LspService::build(..).custom_method("lex/preparePaste", ..)`. Reuses the open buffer's parse; pure w.r.t. document state. Unopened buffer → echoes the clipboard back (safe no-op).
- **Capability**: advertised as `experimental.lexPreparePaste = true` in the initialize response (§5) so editors enable interception only against a server that implements it.

## Decisions / notes

- `mode` is serialized as kebab-case strings (`re-anchor`, `passthrough-verbatim`, …) to match the spec's prose names.
- The transform module lives in `lex-lsp-core` (not `lex-lsp`) because it's pure AST logic — same placement rule the crate already follows for table-navigation etc. The tower-lsp request wiring stays in `lexd-lsp`.
- The `lib.rs` embedding doc-example still shows the minimal `LspService::new`; the binary itself uses `build().custom_method()`. Doc-examples are `doctest = false`, so this is illustrative only.

## Test plan

- `cargo test -p lex-lsp-core -p lexd-lsp` — 28 new tests, all green.
  - 24 in `prepare_paste`: a table-driven sweep of the §6 edge cases (empty clipboard, single-line, selection-replace anchored from selection start, +/0/− delta, fresh-line vs. merge, verbatim + table passthrough, clipboard-that-is-itself-a-verbatim-block stays well-formed, trailing-newline preserved, mixed tabs/spaces, paste at document start, partial indentation) plus classification, anchor-resolution, and wire-format tests.
  - 4 server-level: capability advertisement, re-anchor against an open buffer, verbatim passthrough, unopened-buffer fallback.
- `cargo test --workspace --exclude lex-wasm` — full suite green (38 test binaries, no failures).
- `lefthook run pre-commit --all-files` — green (cargo-fmt, cargo-clippy, yaml/md/shell lint).

Closes #708.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/lex-lsp-core/src/prepare_paste.rs [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+    match line_at(source, pos.line as usize) {
+        Some(line) => {
+            let before: String = line.chars().take(pos.character as usize).collect();
+            before.trim().is_empty()
+        }
+        // No such line (e.g. paste at the very end past the last newline):
+        // there is no pre-existing content to merge into, so it's fresh.
+        None => true,
+    }
+}
```
[the remark this answers is not in the document]
Comment (arthur-debert) [reply crates/lex-lsp-core/src/prepare_paste.rs:307]: Fixed in 6f241ebf. `is_fresh_line` now treats `pos.character` as a UTF-8 byte offset (the server-wide convention — see `slice_text_by_range` in `lexd-lsp`), walking chars and accumulating `len_utf8()` until the caret column, with no intermediate `String` allocation. Added regression tests covering a multi-byte (`café`) line before the caret.

File: crates/lex-lsp-core/src/prepare_paste.rs [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        }
+        let start_line = child.range().start.line;
+        if let Some(line) = line_at(source, start_line) {
+            if !line.trim().is_empty() {
+                return Some(leading_width(line));
+            }
+        }
+    }
+    None
+}
```
[the remark this answers is not in the document]
Comment (arthur-debert) [reply crates/lex-lsp-core/src/prepare_paste.rs:274]: Fixed in 6f241ebf. `resolve_anchor` now pre-splits the source into a `Vec<&str>` once and threads a `&[&str]` slice through `visit_containers` / `content_indent`, replacing the per-child `line_at` full-source re-split. The anchor walk is now linear in document size.
